prov/verbs/src/verbs_info.c

```diff
@@ -982,6 +987,122 @@ static int vrb_ifa_rdma_info(const struct ifaddrs *ifa, char **dev_name,
 	return ret;
 }
 
+int vrb_get_port_space(const struct fi_info *info)
+{
+	if (info != NULL && info->addr_format == FI_SOCKADDR_IB)
+		return RDMA_PS_IB;
+	else
+		return RDMA_PS_TCP;
+}
+
+static int vrb_gid_is_null(const union ibv_gid *gid)
+{
+	return !(gid->raw[8]  | gid->raw[9]  | gid->raw[10] | gid->raw[11] |
+	         gid->raw[12] | gid->raw[13] | gid->raw[14] | gid->raw[15]);
+}
+
+void vrb_set_sid(enum rdma_port_space ps,
+        uint16_t port, struct sockaddr_ib *sib)
+{
+        sib->sib_sid = htonll(((uint64_t) ps << 16) + ntohs(port));
+
+        if (ps)
+                sib->sib_sid_mask = htonll(RDMA_IB_IP_PS_MASK);
+        if (port)
+                sib->sib_sid_mask |= htonll(RDMA_IB_IP_PORT_MASK);
+}
+
+static int vrb_get_rdma_addrinfo(struct ibv_context *context,
+			     uint8_t port_num, int tbl,
+			     struct rdma_addrinfo **rai)
+{
+	union ibv_gid gid;
+	uint16_t pkey;
+	int ret;
+
+	ret = ibv_query_gid(context, port_num, tbl, &gid);
+	if (ret)
+		return -FI_EINVAL;
+
+	if (vrb_gid_is_null(&gid))
```

Comment:
What is this check for?

Reply:
On the system I'm using for development (MOFED stack + IB EDR network adapters), the port attribute `gid_tbl_len` returns multiple GIDs entries, including those that are not set. We need to filter those entries out in order to avoid creating incorrect interfaces.

The function `vrb_gid_is_null()` is inspired by the function `null_gid()` from the utility `ibv_devinfo`:
https://github.com/linux-rdma/rdma-core/blob/master/libibverbs/examples/devinfo.c#L50

Actually, I have changed that function to make it more readable.
```static int vrb_gid_is_null(const union ibv_gid *gid)      
{                                                         
    union ibv_gid null_gid;                               
    memset(&null_gid, 0, sizeof(union ibv_gid));          
    /* Default gid prefix is 0xfe80 */                    
    null_gid.raw[0] = 0xfe;                               
    null_gid.raw[1] = 0x80;                               
                                                          
    return !memcmp(gid, &null_gid, sizeof(union ibv_gid));
}                                                         
```
Please tell me which version for the function do you prefer.